From 5c1af90bf501b6eea29e8775445e694fc7599114 Mon Sep 17 00:00:00 2001
From: Wenqiang Wei <weiwenqiang@mail.ustc.edu.cn>
Date: Thu, 10 Apr 2025 20:19:06 +0800
Subject: [PATCH 1039/1058] LATX, fix: Invalidate checksum_fail_tb.

If the shared memory where TB is located is modified, we should
retranslate it.
---
 accel/tcg/cpu-exec.c      | 10 ++++++++++
 accel/tcg/translate-all.c |  5 +----
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/accel/tcg/cpu-exec.c b/accel/tcg/cpu-exec.c
index 17eab7505f..e8203fc5e0 100644
--- a/accel/tcg/cpu-exec.c
+++ b/accel/tcg/cpu-exec.c
@@ -261,6 +261,16 @@ cpu_tb_exec(CPUState *cpu, TranslationBlock *itb, int *tb_exit)
     env->fpu_clobber = false;
     ret = tcg_qemu_tb_exec(env, tb_ptr);
 
+    if (env->checksum_fail_tb) {
+        TranslationBlock * tb_fail = (TranslationBlock *)env->checksum_fail_tb;
+        lsassert(tb_fail->checksum && tb_fail->pc == env->eip);
+        mmap_lock();
+        tb_phys_invalidate(tb_fail, tb_page_addr0(tb_fail));
+        mmap_unlock();
+        env->checksum_fail_tb = NULL;
+        /*qemu_log("latx checksum fail, retranslate pc=%lx\n", tb_fail->pc);*/
+    }
+
     if (env->insn_save[0]) {
         link_indirect_jmp(env);
     }
diff --git a/accel/tcg/translate-all.c b/accel/tcg/translate-all.c
index 445050b625..0a5aaf4270 100644
--- a/accel/tcg/translate-all.c
+++ b/accel/tcg/translate-all.c
@@ -3523,9 +3523,6 @@ void page_set_flags(target_ulong start, target_ulong end, int flags)
          len != 0;
          len -= TARGET_PAGE_SIZE, addr += TARGET_PAGE_SIZE) {
         PageFlagsNode *p = pageflags_find(addr, addr);
-        if (p && p->flags & PAGE_MEMSHARE) {
-            p->flags |= PAGE_MEMSHARE;
-        }
 #ifdef CONFIG_LATX_AOT
         if (option_aot && (flags & PAGE_WRITE) && p && !(p->flags & PAGE_WRITE)) {
             if (foreach_tb_first(addr, addr + TARGET_PAGE_SIZE)) {
@@ -3542,7 +3539,7 @@ void page_set_flags(target_ulong start, target_ulong end, int flags)
 
     if (flags) {
         inval_tb |= pageflags_set_clear(start, last, flags,
-            ~(reset ? 0 : PAGE_ANON | PAGE_OVERFLOW));
+            ~(reset ? 0 : PAGE_ANON | PAGE_OVERFLOW | PAGE_MEMSHARE));
     }
 
     if (inval_tb) {
-- 
2.49.0

